### PR TITLE
Update DateAdd/DateSub to return correct types.

### PR DESCRIPTION
**Pull request by PavelSafronov**
This change fixes https://github.com/dolthub/dolt/issues/4376

Previous implementation was hard-coded to return sql.Date, but now we are following the MySQL standard and return a type based on the inputs.

The tests in the repo are verifying that the correct data is returned, but we're not testing the actual SQL data that is received, so none of our tests are catching this case yet. We should open a new work item to start testing the actual SQL that's being returned by dolt. For this bug, testing was performed using a local version of dolt with these changes.

The failing query from the original bug is now working:
```
SELECT NOW(), DATE_ADD(NOW(), INTERVAL 14 DAY), dolt_version();
+----------------------------+----------------------------------+----------------+
| NOW()                      | DATE_ADD(NOW(), INTERVAL 14 DAY) | dolt_version() |
+----------------------------+----------------------------------+----------------+
| 2022-10-18 18:53:24.406345 | 2022-11-01 18:53:24.406345       | 0.50.4         |
+----------------------------+----------------------------------+----------------+
1 row in set (0.02 sec)
```

### CODE REVIEW

File: sql/expression/function/date.go [hunk 1]
```diff
@@ -489,3 +496,65 @@ func (c CurrDate) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 func (c CurrDate) WithChildren(children ...sql.Expression) (sql.Expression, error) {
 	return NoArgFuncWithChildren(c, children)
 }
+
+// Determines the return type of a DateAdd/DateSub expression
+// Logic is based on https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-add
+func getType(input sql.Expression, interval *expression.Interval) sql.Type {
```
Comment (zachmu) [inline sql/expression/function/date.go:502]: Keep in mind that packages share a namespace for functions, and this is a large package. And golang doesn't have function overloading.

Suggest something more specific here, like `intervalExprType(...)`

Comment (zachmu) [reply sql/expression/function/date.go:502]: We often omit `get` in functions names in golang

File: sql/expression/function/date.go [hunk 2]
```diff
@@ -489,3 +496,65 @@ func (c CurrDate) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
...
+	inputType := input.Type()
```
Comment (zachmu) [inline sql/expression/function/date.go:503]: If input is nil this is a null pointer exception

**Review by zachmu — COMMENTED** [review]
This is great, but 

1) add engine tests
2) add in handling (and tests) for bind var values (prepared statements)

File: sql/expression/function/date.go (hunk 1 above)
Comment (PavelSafronov) [reply sql/expression/function/date.go:502]: Regarding 1: Engine tests will not catch this bug. I can add the below query to `QueryTests` in [enginetest/queries/queries.go](https://github.com/dolthub/go-mysql-server/blob/main/enginetest/queries/queries.go#L4714-L4717), but that test will pass with or without this fix. The data being returned is correct, we're just assigning it the wrong type. This is another case that would benefit from https://github.com/dolthub/dolt/issues/3646

```
{
  Query:    "SELECT DATE_ADD('2018-05-02 12:13:14', INTERVAL 1 day)",
  Expected: []sql.Row{{time.Date(2018, time.May, 3, 12, 13, 14, 0, time.UTC)}},
}, 
```

Regarding 2: looking into this

Comment (PavelSafronov) [reply sql/expression/function/date.go:502]: Updating.

File: sql/expression/function/date.go (hunk 2 above)
Comment (PavelSafronov) [reply sql/expression/function/date.go:503]: What's the validation approach we should be following? `Eval` is using the same data (`d.Date` and `d.Interval`) without checking them first.
And in general I'm not seeing a lot of nil checks under `sql/expression/function/`, looks like we validate the input data when constructing the objects, then just trust the properties to contain valid data.

Comment (zachmu) [reply sql/expression/function/date.go:506]: Generally speaking every `sql.Expression` can be `nil`, and function implementations need to check for that and do any appropriate null handling

**Review by zachmu — APPROVED** [review]
LGTM!

**Comment by zachmu** [comment]
You have the engine test case covered with the wire format tests.

Another way you could do this is indirectly, via something like CAST, e.g. `select (cast date_add(...) as char)` to make sure that time information is / isn't included.

**Pull request merged**